### PR TITLE
Ensure exceptions in handlers are handled equally for sync and async

### PR DESCRIPTION
**Pull request by fjetter**
I do not have a strong opinion at the moment about what should happen if a handler raises an exception but regardless, I believe an async and sync handler should show the same behaviour. Everything else is pretty hard to distinguish for any developer.

One easy way to achieve this is to simply await the handler during handle_stream. I am not sure if it was implemented this way hoping that messages would be faster worked off or what the motivation behind this was.
If we want to stick to the old `add_callback` we need to implement exception handling on the returned future/task.

### CODE REVIEW

**Comment by fjetter** [comment]
Test failure in `test_no_danglng_asyncio_tasks` is a genuine failure and I can reproduce locally

File: distributed/core.py [hunk 1]
```diff
@@ -594,29 +592,33 @@ async def handle_stream(self, comm, extra=None, every_cycle=[]):
...
             if all(c.cancelled() for c in self._ongoing_coroutines):
                 break
             else:
-                yield asyncio.sleep(0.01)
+                await asyncio.sleep(0.01)
 
         self._event_finished.set()
+        # Allow event listeners a cycle to react before we return
+        # See also https://github.com/dask/distributed/pull/4734
+        await asyncio.sleep(0)
```
Comment (fjetter) [inline distributed/core.py:621]: I received errors with still pending tasks from the worker after shutdown

```
  <Task pending name='Task-41' coro=<Worker.ensure_computing() running at
```

letting the loop run for another cycle here seemed to do the trick but frankly, I don't understand why. Especially not what this has to do with my proposed changes. Is anybody able to shed light on this?

Comment (mrocklin) [reply distributed/core.py:621]: I think that `await asyncio.sleep(0)` may not actually let the event loop cycle.  I think that `sleep(0)` might special-case and continue directly with the same coroutine.  You might want to try something like `sleep(1e-6)` to see if that makes a difference.

Comment (fjetter) [reply distributed/core.py:621]: It is a special case but it does allow one cycle, see https://github.com/python/cpython/blob/accea7dc2bd30a6e8e1b0334acfca9585cbd7f8a/Lib/asyncio/tasks.py#L585-L601 

If this doesn't work I'll try a fixed sleep but this feels fragile

Do you have an idea how this could be connected?

Comment (mrocklin) [reply distributed/core.py:621]: I'm not sure.

Maybe there was a final task submission that previously was ignored but is now actually submitted?

File: distributed/core.py [hunk 2]
```diff
@@ -577,7 +575,7 @@ async def handle_stream(self, comm, extra=None, every_cycle=[]):
 
                 for func in every_cycle:
                     if is_coroutine_function(func):
-                        self.loop.add_callback(func)
```
Comment (fjetter) [inline distributed/core.py:580]: I tracked down the dangling task and this was the `ensure_compute` scheduled here. In particular it concerned the `ensure_computing` which uses the async interface to offload deserialisation. This was introduced in https://github.com/dask/distributed/pull/4307 just before the infamous `2020.12.0` release

**Comment by fjetter** [comment]
This situation is a bit awkward and I see two options

1. We await both handler and every-cycle-callbacks and let asyncio/python deal with the exceptions themselves. This puts us at risk of having a slow handler or slow cycle-callback. This could, for instance, cause the worker to accumulate messages in the queue faster than it could work off. Even worse, I believe there is a chance for the comm / tcp socket to overflow which either results in weird exceptions or dropped messages, I'd need to research this a bit. In either case this would probably only happen if there is a quite significant load on the entire system in which case I'd wonder if we're coming out of this alive, anyhow. Not an expert though.
2. We implement the task offloading to the loop ourselves and take care that the exceptions are properly handled. The way tornado deals with this is not 100% well suited to our use case here. In the cases where we put tasks on the loop, we often would like to check if they completed, erred or cancel them even. Currently we do not have this possibility.

Note: this problem pops up in many places in the worker so solving 2.) once might benefit other situations as well.

@mrocklin Any thoughts on this? Do we have any asyncio/tornado experts we can loop into this discussion?

---

Note: The PR grew again in scope. I ignored the warning in the gen.coroutine comment and figured it was easy to rewrite. Turns out this introduces race conditions during closing since the behaviour of gen.coroutine and plain async coroutines differ in the way they are cancelled (or not). Will break this off into another PR

**Comment by mrocklin** [comment]
> This could, for instance, cause the worker to accumulate messages in the queue faster than it could work off

I'm not sure that this is true.  There is a handle_stream coroutine for every comm.  Comms are most often used with the `rpc` objects as follows:

```python
## caller side
await worker.some_handler(arg=foo)
```

The caller side generally waits until the call to the handler completes before moving on, so the backup won't be inside the socket on the receiver side, it will be in application code on the caller side, which seems safe.  Previously this call would have completed very quickly and now it will take longer.  That's probably ok.

In summary, if things aren't breaking here I think that we're ok.  In principle this change feels safe to me.  I wouldn't be surprised if it did break some things, but even if it did, those things were probably broken beforehand and relying on bad logic in the worker.  For example, if the  caller side was doing the following:

```python
await worker.some_handler(arg=1)
await worker.some_handler(arg=2)
await worker.some_handler(arg=3)
await worker.some_handler(arg=4)
```

Then with async handlers this would have previously run much more quickly than it should have.  In a situation like the following:

```python
await asyncio.gather(
    worker.some_handler(arg=1),
    worker.some_handler(arg=2),
    ...
)
```

The caller side will actually open many comms to the server to handle the requests in parallel, and so we'll get many `handle_stream` coroutines, and everything should behave nicely.

(Usual disclaimer of "I'm a bit rusty on all of this, so please don't believe me")

**Comment by fjetter** [comment]
> I'm not sure that this is true. There is a handle_stream coroutine for every comm. Comms are most often used with the rpc objects as follows:

I'm mostly concerned about the primary link between scheduler and worker which is based on a single comm. Most of the communication between the two is done via this single comm using a `BatchedSend`. For instance all "compute-task", "forget-task" msgs are moving along this single comm and I'm wondering if this becomes an issue.

However, with the exception of `Worker.close` all of our stream handlers seem to be sync anyhow, so I'm worrying unnecessarily 🤦

**Comment by mrocklin** [comment]
> However, with the exception of Worker.close all of our stream handlers seem to be sync anyhow, so I'm worrying unnecessarily

Yeah, anything sent by a BatchedSend is intended to run immediately.  

If you wanted to be careful though you could add a read into the `BatchedSend._background_send` method.  This would close the loop and establish back pressure.  Really we would want to allow a few of these to be in flight before awaiting.  I'm not too worried about this though.

File: distributed/core.py [hunk 3]
```diff
@@ -422,7 +430,7 @@ async def handle_comm(self, comm):
 
         logger.debug("Connection from %r to %s", address, type(self).__name__)
         self._comms[comm] = op
-        await self
```
Comment (fjetter) [inline distributed/core.py:425]: This `await self` without the guard I introduced in `__await__` would allow an incoming comm to restart a currently shutting down scheduler. There is a time window while the scheduler is handling its affairs, before the actual server closes, where it still accepts incoming comms. If during this time a new comm is opened, the scheduler is restarted although it is currently trying to shutdown.

https://github.com/dask/distributed/blob/d9bc3c634a4a3983cdc75e5c39cbdac2afbf6184/distributed/scheduler.py#L3759-L3808

Comment (fjetter) [reply distributed/core.py:425]: I didn't add a dedicated test for this since there are actually many failing tests once the `close` coroutines are shielded

File: distributed/utils.py [hunk 4]
```diff
@@ -202,6 +203,22 @@ def get_ip_interface(ifname):
...
+    function is invoked.
+
+    See also https://docs.python.org/3/library/asyncio-task.html#asyncio.shield
+    """
+
+    @wraps(func)
+    def _(*args, **kwargs):
+        return asyncio.shield(func(*args, **kwargs))
+
+    return _
```
Comment (fjetter) [inline distributed/utils.py:219]: This is not absolutely required but with this decorator we can define methods which are always shielded (e.g. close) instead of relying on remembering it on every invocation

File: distributed/deploy/spec.py [hunk 5]
```diff
@@ -338,15 +338,8 @@ async def _correct_state_internal(self):
             if to_close:
                 if self.scheduler.status == Status.running:
                     await self.scheduler_comm.retire_workers(workers=list(to_close))
-                tasks = [
-                    asyncio.create_task(self.workers[w].close())
-                    for w in to_close
-                    if w in self.workers
-                ]
+                tasks = [self.workers[w].close() for w in to_close if w in self.workers]
```
Comment (mrocklin) [inline distributed/deploy/spec.py:341]: Can we call `asyncio.wait` here and avoid the tasks variable entirely?

File: distributed/nanny.py [hunk 6]
```diff
@@ -506,18 +509,15 @@ async def close(self, comm=None, timeout=5, report=None):
...
             await preload.teardown()
 
-        self.stop()
         try:
             if self.process is not None:
                 await self.kill(timeout=timeout)
         except Exception:
             pass
         self.process = None
-        await self.rpc.close()
```
Comment (mrocklin) [inline distributed/nanny.py:516]: Why this change?

File: distributed/nanny.py [hunk 7]
```diff
@@ -506,18 +509,15 @@ async def close(self, comm=None, timeout=5, report=None):
         for preload in self.preloads:
             await preload.teardown()
 
-        self.stop()
```
Comment (mrocklin) [inline distributed/nanny.py:509]: Why this change?

File: distributed/scheduler.py [hunk 8]
```diff
@@ -3808,10 +3809,6 @@ async def close(self, comm=None, fast=False, close_workers=False):
         for comm in self.client_comms.values():
             comm.abort()
 
-        await self.rpc.close()
-
-        self.status = Status.closed
-        self.stop()
```
Comment (mrocklin) [inline distributed/scheduler.py:3814]: Why this change?

File: distributed/scheduler.py [hunk 9]
```diff
@@ -6257,7 +6259,10 @@ async def retire_workers(
                     )
                 if remove:
                     await asyncio.gather(
-                        *[self.remove_worker(address=w, safe=True) for w in worker_keys]
+                        *[
+                            self.remove_worker(address=w, safe=True, close=False)
+                            for w in worker_keys
+                        ]
```
Comment (mrocklin) [inline distributed/scheduler.py:6265]: Why this change?

**Review by mrocklin — COMMENTED** [review]
I'm curious about the reason for a few of the changes here.  Also, do we know why tests aren't happy yet?

File: distributed/deploy/spec.py (hunk 5 above)
Comment (fjetter) [reply distributed/deploy/spec.py:341]: sure

File: distributed/nanny.py (hunk 7 above)
Comment (fjetter) [reply distributed/nanny.py:509]: It is redundant since we do the same in `Server.close`. it looked like being redundant and I felt removing it would reduce complexity. If there is some purpose to stopping earlier, I can revert this logic. This is only my sense of _cleaning up_

File: distributed/scheduler.py (hunk 8 above)
Comment (fjetter) [reply distributed/scheduler.py:3814]: this is performed by `super().close()` since there are no other things happening in between, letting the server close itself felt like the right thing to do to reduce complexity.

File: distributed/nanny.py (hunk 6 above)
Comment (fjetter) [reply distributed/nanny.py:516]: same argument as above about cleaning up

File: distributed/scheduler.py (hunk 9 above)
Comment (fjetter) [reply distributed/scheduler.py:6265]: This change adds `close=False` (default is True). This causes in most situations to issue a close_worker call twice since there is a dedicated close_worker call above. 

It _might_ be the correct thing to change this to `close=not close_workers` to avoid the duplication

**Comment by jrbourbeau** [comment]
Sorry @fjetter, it looks like https://github.com/dask/distributed/pull/4966 caused some merge conflicts here

**Comment by fjetter** [comment]
I'm hitting a lot of known "flaky" tests here (cannot reproduce locally, yet)

File: distributed/deploy/tests/test_spec_cluster.py [hunk 10]
```diff
@@ -416,6 +415,9 @@ def __str__(self):
 
     __repr__ = __str__
 
+    def __await__(self):
+        return self.start().__await__()
+
```
Comment (fjetter) [inline distributed/deploy/tests/test_spec_cluster.py:420]: This `MultiWorker` is a bit strange since it isn't a Server although `Worker`s are always `Server`s
